### PR TITLE
[WIP] piloting parameter refactor

### PR DESCRIPTION
**Pull request by odunbar**
## Purpose
To prototype the new parameter-from-file interface with Thermodynamics from discussions in #82 

!! Now depends on `https://github.com/CliMA/CLIMAParameters.jl#orad/refactor-parameters-v2` !!
- `add https://github.com/CliMA/CLIMAParameters.jl#orad/refactor-parameters-v2` in `test/Project.toml` (see PR [62](https://github.com/CliMA/CLIMAParameters.jl#orad/refactor-parameters-v2) )
## In this PR 
- Use the default parameter toml file in `orad/refactor-parameters-v2` . Prototype parameters are:
```TOML
 [descriptive_name]
 alias = "deprecated_short_name"
 value = 0.435
 type = "float"
 description = "where necessary this is added"
 ```
- use file parser utilities from `CLIMAParameters`, store parameters in a `Dict` as `alias => value`. 
- create a local parameter struct, pass this to other relevant modules/tests in place of CLIMAParameters
- call parameters in file by `param_set.alias` replacing `CPP.alias(param_set)`
### ALSO: 
- added TOML to `Project.toml` in `test/` 
- added compat of `0.4` to CLIMAParameters in gpuenv
- added `PackageSpec(name="CLIMAParameters",rev="orad/refactor-parameters-v2")` to `.buildkite/pipeline.yml`
## To do  
- currently all naming is done by `alias`  (i.e `param_set.alias`) until the descriptive names are set
- enforcing concrete Float64 type for tests
- All tests pass, except docs (will update last)

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -2,5 +2,19 @@ if !("." in LOAD_PATH) # for easier local testing
     push!(LOAD_PATH, ".")
 end
 
+# read parameters needed for tests
+using TOML
+planet_parse = TOML.parsefile("planet_parameters.toml")
```
Comment (charleskawczynski) [inline test/runtests.jl:7]:
```suggestion
planet_parse = TOML.parsefile(joinpath(@__DIR__, "planet_parameters.toml"))
```

File: test/parameter_tests.jl [hunk 2]
```diff
@@ -0,0 +1,61 @@
+
+
+using Test
+
+
+### read parameters needed for tests
+using TOML
+planet_parse = TOML.parsefile("planet_parameters.toml")
```
Comment (charleskawczynski) [inline test/parameter_tests.jl:8]:
```suggestion
planet_parse = TOML.parsefile(joinpath(@__DIR__, "planet_parameters.toml"))
```

File: perf/alloc_per_constructor.jl [hunk 3]
```diff
@@ -3,6 +3,18 @@ import Pkg
...
+planet_parse = TOML.parsefile(joinpath("test", "planet_parameters.toml"))
+param_dict = Dict{Symbol, Any}()
+for (key, val) in planet_parse
+    # In the future - we will use the full names,
+    # param_dict[Symbol(key)] = val["value"]
+
+    # for now we use the aliases
+    param_dict[Symbol(val["alias"])] = val["value"]
+end
+full_parameter_set = (; param_dict...)
```
Comment (charleskawczynski) [inline perf/alloc_per_constructor.jl:16]: Maybe this should just go in `common.jl`?

Comment (odunbar) [reply perf/alloc_per_constructor.jl:16]: Yeah i see now that this is required in the benchmark tests etc too

File: src/Thermodynamics.jl [hunk 4]
```diff
@@ -51,12 +51,252 @@ const RS = RootSolvers
...
+    kappa_d
+    gas_constant
+    molmass_dryair
+    molmass_water
+    T_surf_ref
+    T_min_ref
+    grav
+    molmass_ratio
+    T_icenuc
+end
```
Comment (charleskawczynski) [inline src/Thermodynamics.jl:215]:
```suggestion
struct ThermodynamicsParameters{FT}
    T_0::FT
    MSLP::FT
    R_d::FT
    cp_d::FT
    cv_d::FT
    cp_v::FT
    cv_v::FT
    cp_l::FT
    cv_l::FT
    cp_i::FT
    cv_i::FT
    LH_v0::FT
    LH_s0::FT
    LH_f0::FT
    e_int_v0::FT
    e_int_i0::FT
    press_triple::FT
    T_triple::FT
    R_v::FT
    T_freeze::FT
    T_min::FT
    T_max::FT
    entropy_reference_temperature::FT
    entropy_dry_air::FT
    entropy_water_vapor::FT
    kappa_d::FT
    gas_constant::FT
    molmass_dryair::FT
    molmass_water::FT
    T_surf_ref::FT
    T_min_ref::FT
    grav::FT
    molmass_ratio::FT
    T_icenuc::FT
end
```

Comment (odunbar) [reply src/Thermodynamics.jl]: done!

File: src/isentropic.jl [hunk 5]
```diff
@@ -45,13 +45,13 @@ The air pressure for an isentropic process, where
...
-    param_set::APS,
+    param_set::ThermodynamicsParameters,
     T::FT,
     T∞::FT,
     p∞::FT,
     ::DryAdiabaticProcess,
 ) where {FT <: AbstractFloat}
-    _kappa_d::FT = CPP.kappa_d(param_set)
+    _kappa_d::FT = param_set.kappa_d
     return p∞ * (T / T∞)^(FT(1) / _kappa_d)
```
Comment (simonbyrne) [inline src/isentropic.jl:55]:
```suggestion
    kappa_d = param_set.kappa_d
    return p∞ * (T / T∞)^(FT(1) / kappa_d)
```

Comment (simonbyrne) [reply src/isentropic.jl:55]: On Julia 1.7, you can also do:
```suggestion
    (;kappa_d,) = param_set
    return p∞ * (T / T∞)^(FT(1) / kappa_d)
```

File: src/isentropic.jl [hunk 6]
```diff
@@ -65,13 +65,13 @@ The air temperature for an isentropic process, where
...
     p::FT,
     θ::FT,
     ::DryAdiabaticProcess,
 ) where {FT <: AbstractFloat}
-    _R_d::FT = CPP.R_d(param_set)
-    _cp_d::FT = CPP.cp_d(param_set)
-    MSLP::FT = CPP.MSLP(param_set)
+    _R_d::FT = param_set.R_d
+    _cp_d::FT = param_set.cp_d
+    MSLP::FT = param_set.MSLP
```
Comment (simonbyrne) [inline src/isentropic.jl:75]:
```suggestion
    (;R_d, cp_d, MSLP) = param_set
```

Comment (simonbyrne) [reply src/isentropic.jl:75]: (again, as an example: we can leave this for now though)

File: src/Thermodynamics.jl [hunk 7]
```diff
@@ -51,12 +51,252 @@ const RS = RootSolvers
...
+    molmass_dryair::FT
+    molmass_water::FT
+    T_surf_ref::FT
+    T_min_ref::FT
+    grav::FT
+    molmass_ratio::FT
+    T_icenuc::FT
+end
+
+function ThermodynamicsParameters(param_set::NamedTuple)
```
Comment (simonbyrne) [inline src/Thermodynamics.jl:217]: I would suggest switching to using a dictionary, rather than a named tuple.

File: .buildkite/pipeline.yml [hunk 8]
```diff
@@ -9,16 +9,14 @@ steps:
     key: "init_cpu_env"
     command:
       - "julia --project=test -e 'using Pkg; Pkg.develop(path=\".\")'"
-      - "julia --project=test -e 'using Pkg; Pkg.rm(\"CLIMAParameters\")'"
-      - "julia --project=test -e 'using Pkg; Pkg.add(url=\"https://github.com/CliMA/CLIMAParameters.jl.git#orad/refactor-parameters-v2\")'"
+      - "julia --project=test -e 'using Pkg; Pkg.add(PackageSpec(\"CLIMAParameters\", rev=\"orad/refactor-parameters-v2\")'"
```
Comment (charleskawczynski) [inline .buildkite/pipeline.yml:12]:
```suggestion
      - "julia --project=test -e 'using Pkg; Pkg.add(PackageSpec(; name=\"CLIMAParameters\", rev=\"orad/refactor-parameters-v2\")'"
```
If this doesn't work, perhaps try this

**Comment by odunbar** [comment]
readdressing in #100

**Pull request closed without merging**